src/geodesic.rs

```diff
@@ -2061,6 +2096,475 @@ mod tests {
+        assert_approx_eq!(s12, 20003898.214, 0.5e-3);
+        let s12: f64 = geod.inverse(89.333123580033, 0.0, -89.333123580032997687, 179.99295812360148422);
+        assert_approx_eq!(s12, 20003926.881, 0.5e-3);
+    }
+
+    #[test]
+    fn test_std_geodesic_geodsolve9() {
+        // Check fix for volatile x bug found 2011-06-25 (gcc 4.4.4 x86 -O3)
+        let geod = Geodesic::wgs84();
+        let s12: f64 = geod.inverse(56.320923501171, 0.0, -56.320923501171, 179.664747671772880215);
+        assert_approx_eq!(s12, 19993558.287, 0.5e-3);
+    }
+
+    #[test]
+    fn test_std_geodesic_geodsolve10() {
+        // Check fix for adjust tol1_ bug found 2011-06-25 (Visual Studio
+        // 10 rel + debug)
+        let geod = Geodesic::wgs84();
+        let s12: f64 = geod.inverse(52.784459512564, 0.0, -52.784459512563990912, 179.634407464943777557);
+        assert_approx_eq!(s12, 19991596.095, 0.5e-3);
+    }
+
+    #[test]
+    fn test_std_geodesic_geodsolve11() {
+        // Check fix for bet2 = -bet1 bug found 2011-06-25 (Visual Studio
+        // 10 rel + debug)
+        let geod = Geodesic::wgs84();
+        let s12: f64 = geod.inverse(48.522876735459, 0.0, -48.52287673545898293, 179.599720456223079643);
+        assert_approx_eq!(s12, 19989144.774, 0.5e-3);
+    }
+
+    #[test]
+    fn test_std_geodesic_geodsolve12() {
+        // Check fix for inverse geodesics on extreme prolate/oblate
+        // ellipsoids Reported 2012-08-29 Stefan Guenther
+        // <stefan.gunther@embl.de>; fixed 2012-10-07
+        let geod = Geodesic::new(89.8, -1.83);
+        let (s12, azi1, azi2, _a12) = geod.inverse(0.0, 0.0, -10.0, 160.0);
+        assert_approx_eq!(azi1, 120.27, 1e-2);
+        assert_approx_eq!(azi2, 105.15, 1e-2);
+        assert_approx_eq!(s12, 266.7, 1e-1);
+    }
+
+    #[test]
+    fn test_std_geodesic_geodsolve14() {
+        // Check fix for inverse ignoring lon12 = nan
+        let geod = Geodesic::wgs84();
+        let (s12, azi1, azi2, _a12) = geod.inverse(0.0, 0.0, 1.0, f64::NAN);
+        assert!(azi1.is_nan());
+        assert!(azi2.is_nan());
+        assert!(s12.is_nan());
+    }
+
+    #[test]
+    #[ignore] // Fails existing behavior.
+    fn test_std_geodesic_geodsolve15() {
+        // Initial implementation of Math::eatanhe was wrong for e^2 < 0.  This
+        // checks that this is fixed.
+        let geod = Geodesic::new(6.4e6, -1f64/150.0);
+        let (_lat2, _lon2, _azi2, _m12, _M12, _M21, S12, _a12) = geod.direct(1.0, 2.0, 3.0, 4.0);
+        assert_approx_eq!(S12, 23700.0, 0.5);
+    }
+
+    #[test]
+    fn test_std_geodesic_geodsolve17() {
+        // Check fix for LONG_UNROLL bug found on 2015-05-07
+        let geod = Geodesic::new(6.4e6, -1f64/150.0);
+        let (_a12, lat2, lon2, azi2, _s12, _m12, _M12, _M21, _S12) =
+            geod._gen_direct(40.0, -75.0, -10.0, false, 2e7, caps::STANDARD | caps::LONG_UNROLL);
+        assert_approx_eq!(lat2, -39.0, 1.0);
```

Comment:
These epsilons seem surprisingly large. Is that copied from the tests? Or is this a situation where geographiclib-rs has a bug and this wide tolerance is trying to accommodate that bug?

Reply:
Those tolerance values are copied from the original. I was also a bit surprised by how relaxed some of the tests are with tolerance, while others are clearly tuned to be as close as possible while using only 1 significant digit in the tolerance. I looked through all the different variations on geodsolve17 just now, and they all use that same relaxed tolerance. I could explore tightening it at some point, but my sense is that we've got higher priority things to look at for a while.